pandas/core/indexes/multi.py

```diff
@@ -2464,6 +2464,7 @@ def get_indexer(self, target, method=None, limit=None, tolerance=None):
         else:
             indexer = self._engine.get_indexer(target)
 
+        self._engine.clear_mapping()
```

Comment:
Why is this being done in `get_indexer`? This seems like the wrong place.

Reply:
The call to self._engine.get_indexer in the lines prior is what causes the memory to blow up so things like joins with multiindexes will also see the same issue since it calls get_indexer.  

```
In [1]: idx = pd.MultiIndex.from_product([pd.date_range('2010-01-01', '2015-01-01'), range(1000)], names=['date', 'id'])
   ...: idx2 = pd.MultiIndex.from_product([pd.date_range('2010-01-01', '2015-01-01'), range(500)], names=['date', 'id'])
   ...: df =  pd.DataFrame({'a': 1}, index=idx)

In [2]: df.memory_usage(deep=True, index=True)
Out[2]:
Index     7453600
a        14616000
dtype: int64

In [3]: df.join(pd.DataFrame(index=idx2))
In [4]: df.memory_usage(deep=True, index=True)
Out[4]:
Index    91339680
a        14616000
dtype: int64

```
Since this is a problem that seems specific to multiindexes and this is the only place that calls the get_indexer method of the MultiIndex engine, I thought this was the best place to address the issue. Lemme know if you think there's a better location though!

Reply:
It's not clear to me that we want to do this though. Allocating the engine can be expensive so this will hurt the performance of things that would reuse the cached engine.